src/hash/hasher.cr

```diff
@@ -0,0 +1,225 @@
+require "crystal/system/random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+# ```
+# class MyHasher
+#   # Value should implement commutative `+` for `Hash#hash(hasher)`
+#   alias Value
+#
+#   # must be implemented to mix sizes of collections, and pointers (object_id)
+#   def raw(v : Int::Primitive)
+#     # mutate
+#     nil
+#   end
+#
+#   # must be implemented for Hash#hash
+#   def raw(v : Value)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(b : Bytes)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(n : Nil)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(v)
+#     # v.hash will return hasher
+#     # if hasher is a struct, then it will be copy
+#     copy_from v.hash(self)
+#     nil
+#   end
+#
+#   # digest returns hashsum for current state without state mutation
+#   def digest : Value
+#   end
+#
+#   # should be implemented for `Hash#hash(hasher)`
+#   def clone
+#     copy_of_current_state
+#   end
+# end
+# ```
+
+class Hash
+  # Hasher used as standard hasher in `Object#hash`
+  # It have to provide defense against HashDos, and to be reasonably fast.
+  #
+  # To protect against HashDos, it is seeded with secure random, and have
+  # permutation that hard to forge without knowing seed and seeing hash digest.
+  struct Hasher
```

Comment:
@funny-falcon I'm somewhat sceptical that LLVM is the issue here, have you tried compiling your funny_hash implementation with clang and comparing to GCC?